Create Cloud Run deployment branch

petechd opened the pull request:
### What is the context of this PR?
This branch is used by new Cloud Run test pipeline in [this PR](https://github.com/ONSdigital/eq-pipelines/pull/241).

### How to review 
Check if script and yaml file changes are all correct and variables are refactored so there are no hardcoded left.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

petechd commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Pipeline a Cloud Run environment (3 SP)](https://trello.com/c/CjGGHDpw/4764-pipeline-a-cloud-run-environment-3-sp)

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+CDN_URL="${CDN_URL:=https://cdn.census.gov.uk}"
+NEW_RELIC_APP_NAME="${NEW_RELIC_APP_NAME:=}"
+NEW_RELIC_LICENSE_KEY="${NEW_RELIC_LICENSE_KEY:=}"
+COOKIE_SETTINGS_URL="${COOKIE_SETTINGS_URL:=#}"
+ADDRESS_LOOKUP_API_URL="${ADDRESS_LOOKUP_API_URL:=#}"
```
I'm not sure these should default to `#`, runner enables these things based on the variables being set.

petechd replied:
I removed these values so it defaults to empty now.

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+    --set-env-vars WEB_SERVER_UWSGI_ASYNC_CORES="${WEB_SERVER_UWSGI_ASYNC_CORES}" \
+    --set-env-vars DATASTORE_USE_GRPC="${DATASTORE_USE_GRPC}" \
+    --set-env-vars HTTP_KEEP_ALIVE="${HTTP_KEEP_ALIVE}" \
+    --set-env-vars EQ_GCS_SUBMISSION_BUCKET_ID="${PROJECT_ID}-submission" \
+    --set-env-vars EQ_GCS_FEEDBACK_BUCKET_ID="${PROJECT_ID}-feedback" \
```
These values should be set in the first block of variables and referenced here as it's a bit hidden here in the list of env vars

petechd replied:
Moved to first block of variables now!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+    --set-env-vars NEW_RELIC_APP_NAME="${NEW_RELIC_APP_NAME}" \
+    --set-env-vars CONFIRMATION_EMAIL_LIMIT="${CONFIRMATION_EMAIL_LIMIT}" \
+    --set-env-vars EQ_SUBMISSION_CONFIRMATION_BACKEND="${EQ_SUBMISSION_CONFIRMATION_BACKEND}" \
+    --vpc-connector="redis-vpc" \
+    --service-account="cloud-run@${PROJECT_ID}.iam.gserviceaccount.com"
```
These two lines are easy to miss here, it would be clearer if they came between the `--image` and `--set-secrets` lines.

petechd replied:
Moved them where requested, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+MEMORY="${MEMORY:=4G}"
+CONCURRENCY="${CONCURRENCY:=80}"
+CDN_URL="${CDN_URL:=https://cdn.census.gov.uk}"
+NEW_RELIC_APP_NAME="${NEW_RELIC_APP_NAME:=}"
```
The ordering of the variables should be consistent between the definition and where they get used in the cloud run command.

petechd replied:
Consistent now, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+ADDRESS_LOOKUP_API_URL="${ADDRESS_LOOKUP_API_URL:=#}"
+ADDRESS_LOOKUP_API_AUTH_ENABLED="${ADDRESS_LOOKUP_API_AUTH_ENABLED:=False}"
+ADDRESS_LOOKUP_API_AUTH_TOKEN_LEEWAY_IN_SECONDS="${ADDRESS_LOOKUP_API_AUTH_TOKEN_LEEWAY_IN_SECONDS:=10}"
+WEB_SERVER_TYPE="${WEB_SERVER_TYPE:=gunicorn-threads}"
+WEB_SERVER_WORKERS="${WEB_SERVER_WORKERS:=7}"
+WEB_SERVER_THREADS="${WEB_SERVER_THREADS:=10}"
+WEB_SERVER_UWSGI_ASYNC_CORES="${WEB_SERVER_UWSGI_ASYNC_CORES:=10}"
```
These four `WEB_SERVER_` variables and `HTTP_KEEP_ALIVE` should be defined before runners environment variables.

petechd replied:
Moved them to the top, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+EQ_GOOGLE_TAG_MANAGER_AUTH="${EQ_GOOGLE_TAG_MANAGER_AUTH:=}"
+CDN_ASSETS_PATH="${CDN_ASSETS_PATH:=/design-system}"
+CONFIRMATION_EMAIL_LIMIT="${CONFIRMATION_EMAIL_LIMIT:=10}"
+EQ_SUBMISSION_CONFIRMATION_BACKEND="${EQ_SUBMISSION_CONFIRMATION_BACKEND:=cloud-tasks}"
```
I think it would be clearer if all of the `_BACKEND` variables were grouped together.

petechd replied:
Grouped them together.

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+WEB_SERVER_THREADS="${WEB_SERVER_THREADS:=10}"
+WEB_SERVER_UWSGI_ASYNC_CORES="${WEB_SERVER_UWSGI_ASYNC_CORES:=10}"
+DATASTORE_USE_GRPC="${DATASTORE_USE_GRPC:=True}"
+MIN_INSTANCES="${MIN_INSTANCES:=1}"
+MAX_INSTANCES="${MAX_INSTANCES:=1}"
```
These should be moved to just after `CONCURRENCY` as they relate to cloud run parameters rather than runner.

petechd replied:
Moved, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE="${EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE:=2021-04-28T02:00:00+00:00}"
+EQ_FEEDBACK_LIMIT="${EQ_FEEDBACK_LIMIT:=10}"
+HTTP_KEEP_ALIVE="${HTTP_KEEP_ALIVE:=650}"
+EQ_GOOGLE_TAG_MANAGER_ID="${EQ_GOOGLE_TAG_MANAGER_ID:=}"
+EQ_GOOGLE_TAG_MANAGER_AUTH="${EQ_GOOGLE_TAG_MANAGER_AUTH:=}"
```
I think these and the New Relic variables should be grouped together as they are external services that runner uses.

petechd replied:
They are grouped together now and separated. Thanks!

petechd commented on app/setup.py:
```diff
@@ -264,7 +264,7 @@ def setup_secure_headers(application):
         csp_policy["connect-src"] += ["ws://localhost:35729"]
 
-    application.config["SESSION_COOKIE_SAMESITE"] = "Strict"
+    application.config["SESSION_COOKIE_SAMESITE"] = "Lax"
```
I think it's one of the requirements on the card: "Ensure samesite cookie policy change (to Lax) is in place"

MebinAbraham replied:
FYI, `Strict` does technically work now that we have custom domains. Lax cookies are sent with cross-site requests, and Strict cookies are not. So `Strict` when launching from the same parent domain works and works as intended, but means we can't launch from non-matching domains like the default Cloud Run URL or localhost since parent domains are different. So it is more to work out how `Strict` worked when on k8s with localhost etc when it shouldn't have?.

So technically for envs maybe apart from `Staging` should be `Strict`. I think card stated that, so it can be investigated separately. We could set it to Strict now and accept env based functional tests won't work for now until we investigate or make it configurable. I think it is appropriate to progress this as `Strict` and investigate separately about launching from localhost etc. 

Thoughts?

ajmaddaford replied:
I wasn't at that tech session. No worries, we can discuss it at the next one.

petechd commented on ci/deploy_credentials.sh:
```diff
@@ -0,0 +1,18 @@
+if [[ -z "$EQ_SECRETS_FILE" ]]; then
+  echo "EQ_SECRETS_FILE not provided"
+  exit 1
+fi
```
Added, thanks!

petechd commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,97 @@
+MAX_INSTANCES="${MAX_INSTANCES:=1}"
+EQ_NEW_RELIC_ENABLED="${EQ_NEW_RELIC_ENABLED:=False}"
+EQ_SECRETS_FILE="${EQ_SECRETS_FILE:=/secrets/secrets.yml}"
+EQ_KEYS_FILE="${EQ_KEYS_FILE:=/keys/keys.yml}"
```
Ok, hardcoded now and variables removed, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,101 @@
+WEB_SERVER_WORKERS="${WEB_SERVER_WORKERS:=7}"
+WEB_SERVER_THREADS="${WEB_SERVER_THREADS:=10}"
+WEB_SERVER_UWSGI_ASYNC_CORES="${WEB_SERVER_UWSGI_ASYNC_CORES:=10}"
+DATASTORE_USE_GRPC="${DATASTORE_USE_GRPC:=True}"
```
This is a runner environment variable, so should be in the next group of variables.

petechd replied:
Moved, thanks!

ajmaddaford commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,101 @@
+    --service-account="cloud-run@${PROJECT_ID}.iam.gserviceaccount.com" \
+    --set-secrets EQ_REDIS_HOST="redis-host:latest" \
+    --set-secrets EQ_REDIS_PORT="redis-port:latest" \
+    --set-secrets "/keys/keys.yml"="keys:latest" \
+    --set-secrets "/secrets/secrets.yml"="secrets:latest" \
+    --set-env-vars EQ_STORAGE_BACKEND="${EQ_STORAGE_BACKEND}" \
+    --set-env-vars EQ_ENABLE_SECURE_SESSION_COOKIE="${EQ_ENABLE_SECURE_SESSION_COOKIE}" \
+    --set-env-vars EQ_SECRETS_FILE="/secrets/secrets.yml" \
+    --set-env-vars EQ_KEYS_FILE="/keys/keys.yml" \
```
The keys and secrets file paths should be defined as variables above, something like:
```
EQ_KEYS_FILE="/keys/keys.yml"
EQ_SECRETS_FILE="/secrets/secrets.yml"
```

and referenced here:
```
--set-secrets "${EQ_KEYS_FILE}"="keys:latest" \
--set-secrets "${EQ_SECRETS_FILE}"="secrets:latest" \
--set-env-vars EQ_KEYS_FILE="${EQ_KEYS_FILE}" \
--set-env-vars EQ_SECRETS_FILE="${EQ_SECRETS_FILE}" \
```

The env vars should also be straight after the `set-secrets`.

petechd replied:
Added now, thanks!

ajmaddaford reviewed (commented):
The main repo Readme needs updating as it still contains kubernetes instructions.

ajmaddaford commented on ci/deploy_app.yaml:
```diff
@@ -2,37 +2,33 @@ platform: linux
   source:
-    repository: ((image_registry))/eq-app-deploy-image
-    tag: ((deploy_image_version))
+    repository: gcr.io/google.com/cloudsdktool/cloud-sdk
+    tag: latest
```
The ci Readme needs to be updated now that `image_registry` and `deploy_image_version` have been removed.

petechd replied:
Updated readme so it doesn't mention `image_registry` and `deploy_image_version`.

petechd commented:
> The main repo Readme needs updating as it still contains kubernetes instructions.

Updated both Readme files to reflect the changes, added new commands examples and new variables to the list.

ajmaddaford commented on README.md:
```diff
@@ -249,17 +234,17 @@ The following environment variables must be set when deploying the app.
-| ROLLING_UPDATE_MAX_UNAVAILABLE     | 25%     | The maximum number of Pods that can be unavailable during the update process     |
-| ROLLING_UPDATE_MAX_SURGE           | 25%     | The maximum number of Pods that can be created over the desired number of Pods   |
-| TARGET_CPU_UTILIZATION_PERCENTAGE  |         | The average CPU utilization usage before auto scaling applies                    |
+| REGION                             |         |                                                                                   |
+| CONCURRENCY                        |         |                                                                                   |
+| MIN_INSTANCES                      |         |                                                                                   |
+| MAX_INSTANCES                      |         |                                                                                   |
+| CPU                                |         |                                                                                   |
+| MEMORY                             |         |                                                                                   |
```
The defaults and descriptions are missing.

ajmaddaford replied:
Also the existing variables need to be checked for correctness e.g. the default for `WEB_SERVER_TYPE` is `gunicorn-async` rather than `gunicorn-threads`.

petechd replied:
Added defaults, descriptions and revisited old values to check for changes, thanks!

MebinAbraham commented on README.md:
```diff
@@ -211,33 +211,18 @@ To run the tests against a remote deployment you will need to specify the enviro
 
 For deploying with Concourse see the [CI README](./ci/README.md).
 
-### Deployment with [Helm](https://helm.sh/)
```
As discussed, we should also document how to deploy directly via `gcloud`.

petechd replied:
Added new `deployment with gcloud` section, thanks!

MebinAbraham commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,103 @@
+    --image="${DOCKER_REGISTRY}/eq-questionnaire-runner:${IMAGE_TAG}" --platform=managed --allow-unauthenticated \
+    --vpc-connector="redis-vpc" \
+    --service-account="cloud-run@${PROJECT_ID}.iam.gserviceaccount.com" \
+    --set-env-vars WEB_SERVER_TYPE="${WEB_SERVER_TYPE}" \
+    --set-env-vars WEB_SERVER_WORKERS="${WEB_SERVER_WORKERS}" \
+    --set-env-vars WEB_SERVER_THREADS="${WEB_SERVER_THREADS}" \
+    --set-env-vars WEB_SERVER_UWSGI_ASYNC_CORES="${WEB_SERVER_UWSGI_ASYNC_CORES}" \
+    --set-env-vars HTTP_KEEP_ALIVE="${HTTP_KEEP_ALIVE}" \
+    --set-secrets EQ_REDIS_HOST="redis-host:latest" \
+    --set-secrets EQ_REDIS_PORT="redis-port:latest" \
+    --set-secrets "${EQ_KEYS_FILE}"="keys:latest" \
+    --set-secrets "${EQ_SECRETS_FILE}"="secrets:latest" \
+    --set-env-vars EQ_KEYS_FILE="${EQ_KEYS_FILE}" \
+    --set-env-vars EQ_SECRETS_FILE="${EQ_SECRETS_FILE}" \
```
Minor, but can the secrets be set first, then the env vars?. Looks odd, that they are mixed.

petechd replied:
Done, thanks!

MebinAbraham commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,103 @@
+NEW_RELIC_LICENSE_KEY="${NEW_RELIC_LICENSE_KEY:=}"
+
+
+gcloud beta run deploy eq-questionnaire-runner \
+    --project="${PROJECT_ID}" --region="${REGION}" --concurrency="${CONCURRENCY}" --min-instances="${MIN_INSTANCES}" --max-instances="${MAX_INSTANCES}" \
+    --port=5000 --cpu="${CPU}" --memory="${MEMORY}" \
+    --image="${DOCKER_REGISTRY}/eq-questionnaire-runner:${IMAGE_TAG}" --platform=managed --allow-unauthenticated \
```
All variables in this file that doesn't have a default, needs to be validated, so for example, vars like project_id, docker registry, image tag etc must all exist. 
Same validation as what we have in `deploy_credentials.sh`, i.e:
```sh
if [[ -z "$PROJECT_ID" ]]; then
  echo "PROJECT_ID not provided"
  exit 1
fi
```

petechd replied:
Added validation, thanks!

MebinAbraham commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,103 @@
+EQ_INDIVIDUAL_RESPONSE_LIMIT="${EQ_INDIVIDUAL_RESPONSE_LIMIT:=75}"
+EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE="${EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE:=2021-04-28T02:00:00+00:00}"
+EQ_FEEDBACK_LIMIT="${EQ_FEEDBACK_LIMIT:=10}"
+EQ_GCS_SUBMISSION_BUCKET_ID="${PROJECT_ID}-submission"
+EQ_GCS_FEEDBACK_BUCKET_ID="${PROJECT_ID}-feedback"
```
If values such as this will be hard coded to this format, then we should update the deploy yaml to not pass them and also the pipelines.yaml shouldn't pass it either as it wouldn't get used here.

petechd replied:
So these values are not passed in deploy.yaml and test pipeline.yaml, fixed👍

petechd commented on ci/deploy_app.yaml:
```diff
@@ -2,37 +2,33 @@ platform: linux
   DATASTORE_USE_GRPC:
   CLOUD_ARMOR_POLICY_NAME:
+  MIN_INSTANCES:
+  MAX_INSTANCES:
```
Aligned and added all vars to yaml file, thanks!

MebinAbraham commented on ci/deploy_app.sh:
```diff
@@ -0,0 +1,119 @@
+EQ_INDIVIDUAL_RESPONSE_LIMIT="${EQ_INDIVIDUAL_RESPONSE_LIMIT:=75}"
+EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE="${EQ_INDIVIDUAL_RESPONSE_POSTAL_DEADLINE:=2021-04-28T02:00:00+00:00}"
+EQ_FEEDBACK_LIMIT="${EQ_FEEDBACK_LIMIT:=10}"
+SUBMISSION_BUCKET_NAME="${PROJECT_ID}-submission"
+FEEDBACK_BUCKET_NAME="${PROJECT_ID}-feedback"
```
These should be `EQ_GCS_SUBMISSION_BUCKET_ID` and `EQ_GCS_FEEDBACK_BUCKET_ID`.

MebinAbraham commented on app/setup.py:
```diff
@@ -321,10 +321,10 @@ def setup_redis(application):
-            raise Exception("Setting SUBMISSION_BUCKET_NAME Missing")
+            raise Exception("Setting EQ_SUBMISSION_BUCKET_NAME Missing")
 
         application.eq["submitter"] = GCSSubmitter(bucket_name=bucket_id)
```
The variable should be updated to `bucket_name`.

petechd replied:
Done everywhere, including tests names. Thanks!

MebinAbraham commented on README.md:
```diff
@@ -246,26 +246,26 @@ The following environment variables must be set when deploying the app.
 | Variable Name                             | Description                                                                          |
 |-------------------------------------------|--------------------------------------------------------------------------------------|
-| SUBMISSION_BUCKET_NAME                    | The name of the bucket that submissions will be stored in                            |
+| EQ_SUBMISSION_BUCKET_NAME                 | The name of the bucket that submissions will be stored in                            |
```
This should be removed, it is optional now.

petechd replied:
Removed it from here, thanks!

MebinAbraham commented on ci/README.md:
```diff
@@ -24,12 +24,10 @@ In addition to the environment variables specified in [Deploying the app](../REA
 | REGION                                    | What region to authenticate against                                                  |
 | PROJECT_ID                                | The ID of the GCP target project                                                     |
```
This table of envs vars should be removed as it is a duplicate. It just needs the link to the deploying the app section of the main readme.

petechd replied:
Linked now instead of this table of variables, thanks!

MebinAbraham commented on ci/deploy_credentials.yaml:
```diff
@@ -3,11 +3,11 @@ image_resource:
   type: docker-image
   source:
     repository: gcr.io/google.com/cloudsdktool/cloud-sdk
+    tag: latest
```
For `gcloud secrets` the alpine image is sufficient.
```suggestion
    tag: alpine
```

petechd commented on ci/README.md:
````diff
@@ -1,10 +1,10 @@
+Before deploying the app you need to create credentials. This can be done via Concourse using the following task commands:
 
 ```sh
 REGION=<cluster_region> \
````
Removed, thanks!

petechd commented on README.md:
```diff
@@ -246,26 +246,26 @@ The following environment variables must be set when deploying the app.
 |-------------------------------------------|--------------------------------------------------------------------------------------|
-| SUBMISSION_BUCKET_NAME                    | The name of the bucket that submissions will be stored in                            |
+| EQ_SUBMISSION_BUCKET_NAME                 | The name of the bucket that submissions will be stored in                            |
 | DOCKER_REGISTRY                           | The FQDN of the target Docker registry                                               |
 | IMAGE_TAG                                 |                                                                                      |
```
Did all these changes requested, thanks!
